go/common/subscription/new_heads_manager.go

```diff
@@ -42,38 +57,60 @@ func NewNewHeadsService(inputCh chan *common.BatchHeader, convertToEthHeader boo
 }
 
 func (nhs *NewHeadsService) Start() error {
-	go ForwardFromChannels([]chan *common.BatchHeader{nhs.inputCh}, nhs.stopped, func(head *common.BatchHeader) error {
-		if nhs.onMessage != nil {
-			err := nhs.onMessage(head)
-			if err != nil {
-				nhs.logger.Info("failed invoking onMessage callback.", log.ErrKey, err)
-			}
+	if nhs.inputCh == nil {
+		nhs.inputCh = make(chan *common.BatchHeader)
+		err := nhs.connectFunc(nhs.inputCh)
+		if err != nil {
+			return fmt.Errorf("could not connect to new heads: %w", err)
 		}
+	}
 
-		var msg any = head
-		if nhs.convertToEthHeader {
-			msg = convertBatchHeader(head)
-		}
+	go ForwardFromChannels(
+		[]chan *common.BatchHeader{nhs.inputCh},
+		nhs.stopped,
+		func(head *common.BatchHeader) error {
+			if nhs.onMessage != nil {
+				err := nhs.onMessage(head)
+				if err != nil {
+					nhs.logger.Info("failed invoking onMessage callback.", log.ErrKey, err)
+				}
+			}
 
-		nhs.notifiersMutex.RLock()
-		defer nhs.notifiersMutex.RUnlock()
+			var msg any = head
+			if nhs.convertToEthHeader {
+				msg = convertBatchHeader(head)
+			}
+
+			nhs.notifiersMutex.RLock()
+			defer nhs.notifiersMutex.RUnlock()
 
-		// for each new head, notify all registered subscriptions
-		for id, notifier := range nhs.newHeadNotifiers {
-			if nhs.stopped.Load() {
-				return nil
+			// for each new head, notify all registered subscriptions
+			for id, notifier := range nhs.newHeadNotifiers {
+				if nhs.stopped.Load() {
+					return nil
+				}
+				err := notifier.Notify(id, msg)
+				if err != nil {
+					// on error, remove the notification
+					nhs.logger.Info("failed to notify newHead subscription", log.ErrKey, err, log.SubIDKey, id)
+					nhs.notifiersMutex.Lock()
```

Comment:
This would deadlock I think, you can't take the write-lock while it's under read-lock. Need to copy the notifiers list before iterating or make list of ones to delete after maybe.

Reply:
I haven't changed this bit. Just reformatted. It might have been there before. I'll have a look